src/Compression/Compression.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+using ObjCRuntime;
+
+namespace Compression {
+
+
+	public partial class CompressionStream : IDisposable {
+
+		[StructLayout(LayoutKind.Sequential)]
+		struct CompressionStreamStruct {
+			public IntPtr Destination;
+			public int DestinationLength;
+			public IntPtr Source;
+			public int SourceLength;
+			public IntPtr State;
+		}
+
+		[DllImport (Constants.libcompression)]
+		static extern int /* CompressionStatus */ CompressionStreamInit (ref CompressionStreamStruct stream, int /* StreamOperation */ operation, int /* CompressionAlgorithm */ algorithm);
+
+		[DllImport (Constants.libcompression)]
+		static extern int /* CompressionStatus */ CompressionStreamProcess (ref CompressionStreamStruct stream, int /* StreamFlag */ flags);
+
+		[DllImport (Constants.libcompression)]
+		static extern int /* CompressionStatus */ CompressionStreamDestroy (ref CompressionStreamStruct stream);
+
+		static int defaultInBufSize = 256;
+		static int defaultOutBufSize = 1024;
+
+		int inSize;
+		int outSize;
+		bool disposedValue = false;
+		bool finalizeStream = false;
+		CompressionStatus status;
+		StreamOperation operation;
+		CompressionStreamStruct internalStream;
+		Stream sourceStream;
+		Stream destinationStream;
+		int totalInputSize = 0;
+		int totalOutputSize = 0;
+		IntPtr dstBuf;
+		IntPtr srcBuf;
+
+		public CompressionStream (Stream inSourceStream, Stream inDestinationStream, StreamOperation inOperation, CompressionAlgorithm algorithm)
+			: this (inSourceStream, inDestinationStream, inOperation, algorithm, defaultInBufSize, defaultOutBufSize) { }
+
+		public CompressionStream (Stream inSourceStream, Stream inDestinationStream, StreamOperation inOperation, CompressionAlgorithm algorithm, int inBufferSize, int outBufferSize)
+		{
+			operation = inOperation;
+			sourceStream = inSourceStream;
+			destinationStream = inDestinationStream;
+			status = (CompressionStatus) CompressionStreamInit (ref internalStream, (int) operation, (int) algorithm);
+			if (status != CompressionStatus.Ok)
+				throw new InvalidOperationException ("Internal stream could not be created.");
+
+			// set the different required values
+			inSize = inBufferSize;
+			outSize = outBufferSize;
+			dstBuf = Marshal.AllocHGlobal (outSize);
```

Comment:
`dstBuf` and `srcBuf` are never freed.

Reply:
@rolfbjarne they destroy method in theory does take care of them. Let me double check the docs.